File: packages_py/fetch_client/src/fetch_client/health/status_checker.py

from typing import Any, Optional

from ..health.models import FetchStatus, FetchStatusResult


class FetchStatusChecker:
    """Check fetch connectivity to a provider using runtime config."""

    def __init__(
        self,
        provider_name: str,
        runtime_config: Any,  # ComputeAllResult
        timeout_seconds: float = 10.0,
        endpoint_override: Optional[str] = None,
    ):
        self.provider_name = provider_name
        self.runtime_config = runtime_config
        self.timeout_seconds = timeout_seconds
        self.endpoint_override = endpoint_override

    async def check(self) -> FetchStatusResult:
        """Execute health check and return result."""
        # Defer import to avoid circular dependency
        from ..provider.provider_client import ProviderClient

        # Wrap instantiation in try/except because validation happens in constructor
        try:
            provider = ProviderClient(
                self.provider_name,
                self.runtime_config,
                options=None, # will use defaults or I should construct it
            )
            # Re-set options to match passed args (constructor defaults might differ?)
            # ProviderClient options: timeout_seconds=10.0.
            # We should pass them.
            provider.options.timeout_seconds = self.timeout_seconds
            provider.options.endpoint_override = self.endpoint_override

            try:
                return await provider.check_health()
            finally:
                await provider.close()

        except Exception as e:
            # Maintain backward compatibility
            from datetime import datetime, timezone
            timestamp = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
            
            # Dummy empty objects for response
            empty_req = {"method": "UNKNOWN", "url": "UNKNOWN", "timeout_seconds": 0}
            empty_config = {
                "base_url": getattr(getattr(self.runtime_config, "config", None) or {}, "base_url", ""),
                "health_endpoint": "UNKNOWN",
                "method": "UNKNOWN",
                "timeout_seconds": 0,
                "auth_type": None,
                "auth_resolved": False,
                "auth_header_present": False,
                "is_placeholder": None,
                "proxy_url": None,
                "proxy_resolved": False,
                "headers_count": 0
            }
            empty_opts = {
                "method": "UNKNOWN",
                "url": "UNKNOWN",
                "timeout_seconds": 0,
                "headers": {},
                "headers_count": 0,
                "follow_redirects": False,
                "proxy": None,
                "verify_ssl": False
            }

            return FetchStatusResult(
                provider_name=self.provider_name,
                status=FetchStatus.CONFIG_ERROR,
                latency_ms=0,
                timestamp=timestamp,
                request=empty_req,
                config_used=empty_config,
                fetch_option_used=empty_opts,
                error={"type": "ConfigError", "message": str(e)},
            )

File: packages_mjs/fetch-client/src/health/status-checker.ts

/**
 * Fetch Status Checker Implementation.
 */
import { FetchStatus, FetchStatusResult } from "./models.js";
import { ProviderClient, ComputeAllResult } from "../provider/provider-client.js";

export class FetchStatusChecker {
  constructor(
    private provider_name: string,
    private runtime_config: ComputeAllResult,
    private timeout_seconds: number = 10.0,
    private endpoint_override?: string
  ) { }

  public async check(): Promise<FetchStatusResult> {
    try {
      const provider = new ProviderClient(
        this.provider_name,
        this.runtime_config,
        {
          timeoutSeconds: this.timeout_seconds,
          endpointOverride: this.endpoint_override,
        }
      );

      try {
        return await provider.checkHealth();
      } finally {
        await provider.close();
      }
    } catch (e: any) {
      // console.log('DEBUG: Caught error in FetchStatusChecker:', e);
      // Maintain backward compatibility: validation errors return CONFIG_ERROR status
      // instead of throwing.
      const timestamp = new Date().toISOString();
      const result = {
        provider_name: this.provider_name,
        status: FetchStatus.CONFIG_ERROR,
        latency_ms: 0,
        timestamp,
        request: { method: 'UNKNOWN', url: 'UNKNOWN', timeout_seconds: 0 },
        config_used: {
          base_url: this.runtime_config.config.base_url || '',
          health_endpoint: 'UNKNOWN',
          method: 'UNKNOWN',
          timeout_seconds: 0,
          auth_type: null,
          auth_resolved: false,
          auth_header_present: false,
          is_placeholder: null,
          proxy_url: null,
          proxy_resolved: false,
          headers_count: 0
        },
        fetch_option_used: {
          method: 'UNKNOWN',
          url: 'UNKNOWN',
          timeout_seconds: 0,
          headers: {},
          headers_count: 0,
          follow_redirects: false,
          proxy: null,
          verify_ssl: false
        },
        error: {
          type: 'ConfigError',
          message: e.message || String(e)
        }
      };
      // console.log('DEBUG: Returning error result:', result);
      return result;
    }
  }
}
